cli/src/main.rs

```diff
@@ -37,172 +54,393 @@ struct Args {
+        let port = get_port()?;
+
+        let run_preview_server = || async {
+            let routes =
+                get_server_config(document.clone(), connections.clone(), current_path.clone());
+            warp::serve(routes).run(([127, 0, 0, 1], port)).await
+        };
+
+        let watch_dir = || async {
+            watch_files(
+                Some(document.clone()),
+                Some(connections.clone()),
+                &args,
+                &current_path,
+            )
+            .await
+        };
+
+        println!("started server and watching dir {current_path:?}");
+
+        // TODO LATER: Investigate the performance penality of joining the server with the compiler like this
+        // in comparison to using tokio::task::spawn to create a seperate thread.
+        let (_, watch_result) = tokio::join!(run_preview_server(), watch_dir());
+
+        watch_result?; // Propagate errors from the watcher and compiler
+
+        return Ok(());
+    }
+
+    // using the watch flag but with some other output format
     if args.watch {
-        watch(&args, &target)?;
-    } else {
-        match compile_file(&args) {
-            Ok((tree, state)) => {
-                let mut stdout = stdout();
-
-                stdout.execute(terminal::Clear(terminal::ClearType::All))?;
-                stdout.execute(cursor::MoveTo(0, 0))?;
-                stdout.execute(style::PrintStyledContent(
-                    "File successfully compiled!\n\n".green(),
-                ))?;
-
-                if !state.warnings.is_empty() {
-                    stdout.execute(style::PrintStyledContent("Warnings:\n".yellow()))?;
-                    for warning in state.warnings {
-                        stdout.execute(style::PrintStyledContent(format!("{warning}").yellow()))?;
-                    }
-                }
+        // Just start the file watcher, but without html live preview
+        // which means that we wil provide None instead of the document and connections
+        return watch_files(None, None, &args, &current_path).await;
+    }
 
-                if !state.errors.is_empty() {
-                    stdout.execute(style::PrintStyledContent("Errors:\n".red()))?;
-                    for error in state.errors {
-                        stdout.execute(style::PrintStyledContent(format!("{error}").red()))?;
-                    }
-                }
+    // Otherwise, if they are using the watcher or live preview
+    // just compile the file once and save it.
+    print_compiling_message()?;
+    let compilation_result = compile_file(&args.input, &args.get_output_format()?);
+    save_result(&compilation_result, &args)?;
+    print_result(&compilation_result, &args)?;
+
+    Ok(())
+}
+
+/// Choose a free port for hosting the html live preview
+fn get_port() -> Result<u16, CliError> {
+    PREVIEW_PORT
+        .get_or_init(|| portpicker::pick_unused_port())
+        .ok_or(CliError::NoFreePorts)
```

Comment:
It is nice to have a check if there are free ports available and fail the live server otherwise, however, assuming that there are no free ports (or that we block modmark from creating sockets, which wouldn't be too unreasonable for some users depending on system settings), we can no longer compile to html with the watch flag at all; `modmark --watch in.mdm out.html` exits immediately saying "No free ports" without compiling the document